lang/jvm/src/main/scala/com/wavesplatform/lang/JavaAdapter.scala

```diff
@@ -1,29 +1,171 @@
-import com.wavesplatform.lang.v1.compiler.Terms.EXPR
+import cats.implicits._
+import com.wavesplatform.common.utils.EitherExt2
+import com.wavesplatform.lang
+import com.wavesplatform.lang.ScriptMeta.{FunctionArgument, FunctionSignature}
+import com.wavesplatform.lang.directives.values._
+import com.wavesplatform.lang.directives.{Directive, DirectiveDictionary, DirectiveParser, DirectiveSet}
+import com.wavesplatform.lang.script.v1.ExprScript
+import com.wavesplatform.lang.script.{ContractScript, ScriptReader}
+import com.wavesplatform.lang.v1.CTX
+import com.wavesplatform.lang.v1.compiler.{CompilerContext, ContractCompiler, Decompiler, ExpressionCompiler}
+import com.wavesplatform.lang.v1.estimator.v3.ScriptEstimatorV3
 import com.wavesplatform.lang.v1.evaluator.ctx.impl.waves.WavesContext
 import com.wavesplatform.lang.v1.evaluator.ctx.impl.{CryptoContext, PureContext}
+import com.wavesplatform.lang.v1.repl.node.http.NodeConnectionSettings
+import com.wavesplatform.lang.v1.repl.{Repl, _}
+import com.wavesplatform.lang.v1.traits.Environment
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.jdk.CollectionConverters._
 
 object JavaAdapter {
-  private val version = V2
-
-  lazy val ctx =
-    Monoid.combineAll(
-      Seq(
-        CryptoContext.compilerContext(Global, version),
-        WavesContext.build(???).compilerContext,
-        PureContext.build(version).compilerContext
-      ))
-
-  def compile(input: String): EXPR = {
-    ExpressionCompiler
-      .compile(input, ctx)
-      .fold(
-        error => throw new IllegalArgumentException(error),
-        res => res
-      )
+  private val allDirectives: Iterable[Either[ExecutionError, DirectiveSet]] =
+    for {
+      version     <- DirectiveDictionary[StdLibVersion].all
+      scriptType  <- DirectiveDictionary[ScriptType].all
+      contentType <- DirectiveDictionary[ContentType].all
+    } yield DirectiveSet(version, scriptType, contentType)
+
+  private val ctxCache: Map[DirectiveSet, CTX[Environment]] =
+    allDirectives
+      .filter(_.isRight)
+      .map(_.explicitGet())
+      .map(ds => (ds, buildCtx(ds)))
+      .toMap
+
+  private def buildCtx(directiveSet: DirectiveSet): CTX[Environment] =
+    CryptoContext.build(Global, directiveSet.stdLibVersion).withEnvironment[Environment] |+|
+      WavesContext.build(directiveSet) |+|
+      PureContext.build(directiveSet.stdLibVersion).withEnvironment[Environment]
+
+  def compile(input: String): Script = {
+    val script =
+      for {
+        directives   <- DirectiveParser(input)
+        directiveSet <- Directive.extractDirectives(directives, defaultStdLib = V4)
+        compile = if (directiveSet.contentType == DApp)
+          compileDApp _
+        else
+          compileExpression _
+        compiled <- compile(input, ctxCache(directiveSet).compilerContext, directiveSet)
+      } yield compiled
+
+    script.fold(error => throw new RideException(error), identity)
```

Comment:
maybe handled exception?

Reply:
Checked expcetion?
It is not so useful.